### ISSUE TITLE
performance: `fn:collection` significantly slower in version 6 than version 4 of existdb

### ISSUE THREAD

**Issue by line-o** [eXist-db/exist#4714]
I was asked to do a speed comparison of `fn:collection` between versions 4 and 6 of exist.

- systems under test: 4.11.1 and 6.1.0 (both build from source and running side by side on bare metal on an ARM mac)
- A test run consists of 100 repeated calls to `fn:collection("/db/test")`. Each call is timed individually and the sequence is then aggregated.
- The test collection contained about 70000, small xml files in total.
- The files where in buckets of 50 files per bucket.

### eXist-4.11.1
Time spent in milliseconds

- minimum: 16
- maximum: 27
- average: 17.5
- total: 1750

### eXist-6.1.0
Time spent in milliseconds

- minimum: 40
- maximum: 106
- average: 42.84
- total: 4284

**Expected behavior**

I would not have expected the older version to outperform exist 6 by a factor of 2 (a little over even). We should investigate if we find ways to speed up this core functionality again. This must not trade of stability and correctness,
which I expect are the reason for the slow-down.

**To Reproduce**

```Xquery
xquery version "3.1";

declare variable $local:millisecond := xs:dayTimeDuration('PT0.001S');
declare variable $local:initial-stats :=
    map { 'min': (), 'max': (), 'avg': 0.0, 'sum': 0, 'length': 0 };

declare
function local:sequence-stats-reducer ($result as map(*), $next as xs:numeric) as map(*) {
    map {
        'min': min(($result?min, $next)),
        'max': max(($result?max, $next)),
        'avg': ($result?sum + $next) div ($result?length + 1),
        'sum': $result?sum + $next,
        'length': $result?length + 1
    }
};

declare function local:test ($f as function(*)) {
    let $start := util:system-time()

    let $run := $f()

    return (util:system-time() - $start) div $local:millisecond
};

declare function local:action () {
    (: count( :)
        collection("/db/test")
    (: ) :)
};

let $results := for-each(1 to 100, function ($_) { local:test(local:action#0) })
let $stats := fold-left($results, $local:initial-stats, local:sequence-stats-reducer#2)

return serialize($stats, map{"method": "json"})
```

1. locate or create a collection with data you want to test with
2. save the module to your database or filesystem
3. run the above query in eXide or using `xst run -f`
4. see the aggregated results in JSON

**Context (please always complete the following information):**
- Build: eXist-6.2.0-SNAPSHOT and eXist-4.11.1
- Java: 1.8.0_362 (Azul Systems, Inc.)
- OS: Mac OS X 12.6.2 (aarch64)

**Additional context**
- How is eXist-db installed? built from source
- Any custom changes in e.g. `conf.xml`? none

**Comment by adamretter** [comment]
These results are absolutely inline with what I would expect. eXist-db 6 is slower in areas that require taking Collection and Document locks. Evolved Binary put a vast amount of effort into fixing the locking problems of eXist-db 4 to create eXist-db 5. Locking is expensive, but whereas 4.x.x had many data corruption issues and deadlocks, 6.x.x has very few because Evolved Binary fixed most of the locking issues. These fixes involved adding a lot of missing locks. Stability and correctness come before performance.

**Comment by line-o** [comment]
> Stability and correctness come before performance.

I agree. Since it must have gone unnoticed, I emphasised the relevant sentence in my issue.

**Comment by adamretter** [comment]
> Since it must have gone unnoticed

It did not go unnoticed!
I was simply pointing out the goals that we as Evolved Binary set out to achieve at that time with that piece of work. I believe that also supports your point above.

**Comment by duncdrum** [comment]
@line-o is there something actionable we should undertake from the experiment? If not I propose to close this.
